storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java

```diff
@@ -0,0 +1,420 @@
+    private final Map<String, Object> _conf;
+    private final Map<String, LocalDownloadedResource> _basicPending;
+    private final Map<String, LocalDownloadedResource> _blobPending;
+    private final AdvancedFSOps _fsOps;
+
+    private class DownloadBaseBlobsDistributed implements Callable<Void> {
+        private final String _topologyId;
+        
+        public DownloadBaseBlobsDistributed(String topologyId) {
+            this._topologyId = topologyId;
+        }
+        
+        @Override
+        public Void call() throws Exception {
+            String stormroot = ConfigUtils.supervisorStormDistRoot(_conf, _topologyId);
+            File sr = new File(stormroot);
+            if (sr.exists()) {
+                if (!_fsOps.supportsAtomicDirectoryMove()) {
+                    LOG.warn("{} may have partially downloaded blobs, recovering", _topologyId);
+                    Utils.forceDelete(stormroot);
+                } else {
+                    LOG.warn("{} already downloaded blobs, skipping", _topologyId);
+                    return null;
+                }
+            }
+            boolean deleteAll = true;
+            String tmproot = ConfigUtils.supervisorTmpDir(_conf) + Utils.FILE_PATH_SEPARATOR + Utils.uuid();
+            try {
+                String stormJarKey = ConfigUtils.masterStormJarKey(_topologyId);
+                String stormCodeKey = ConfigUtils.masterStormCodeKey(_topologyId);
+                String stormConfKey = ConfigUtils.masterStormConfKey(_topologyId);
+                String jarPath = ConfigUtils.supervisorStormJarPath(tmproot);
+                String codePath = ConfigUtils.supervisorStormCodePath(tmproot);
+                String confPath = ConfigUtils.supervisorStormConfPath(tmproot);
+                FileUtils.forceMkdir(new File(tmproot));
+                _fsOps.restrictDirectoryPermissions(tmproot);
+                ClientBlobStore blobStore = Utils.getClientBlobStoreForSupervisor(_conf);
+                try {
+                    Utils.downloadResourcesAsSupervisor(stormJarKey, jarPath, blobStore);
+                    Utils.downloadResourcesAsSupervisor(stormCodeKey, codePath, blobStore);
+                    Utils.downloadResourcesAsSupervisor(stormConfKey, confPath, blobStore);
+                } finally {
+                    blobStore.shutdown();
+                }
+                Utils.extractDirFromJar(jarPath, ConfigUtils.RESOURCES_SUBDIR, tmproot);
+                _fsOps.moveDirectoryPreferAtomic(new File(tmproot), new File(stormroot));
+                SupervisorUtils.setupStormCodeDir(_conf, ConfigUtils.readSupervisorStormConf(_conf, _topologyId), stormroot);
+                deleteAll = false;
+            } finally {
+                if (deleteAll) {
+                    LOG.info("Failed to download basic resources for topology-id {}", _topologyId);
+                    Utils.forceDelete(tmproot);
+                    Utils.forceDelete(stormroot);
```

Comment:
We can catch and ignore any IOException from these forceDelete calls so that the original exception bubbles up.